File: backend/app/schemas/topology.py
"""Pydantic schemas for topology APIs."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.topology import NodeType, TopologyStatus


class TopologyCreate(BaseModel):
    """Payload for creating a topology definition."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "name": "edge-demo",
                "description": "Host + service on shared bridge",
                "runtime_target": "docker",
                "networking_mode": "docker_bridge",
                "status": "draft",
                "config": None,
            }
        }
    )

    name: str = Field(
        ...,
        min_length=1,
        max_length=255,
        description="Human-readable name shown in UIs and deployment events.",
    )
    description: str | None = Field(
        default=None,
        description="Optional longer description; not interpreted by the runtime provider.",
    )
    runtime_target: str = Field(
        ...,
        max_length=64,
        description="Runtime key (e.g. `docker`) used by `runtime_provider_for_topology`.",
    )
    networking_mode: str = Field(
        ...,
        max_length=64,
        description="How overlay/bridge networking should be interpreted when planning deploy.",
    )
    status: TopologyStatus | None = Field(
        default=None,
        description="Defaults to draft when omitted.",
    )
    config: dict[str, Any] | None = Field(
        default=None,
        description="Opaque JSON bag for future planner hints.",
    )


class TopologyResponse(BaseModel):
    """Topology row returned from the API."""

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": "550e8400-e29b-41d4-a716-446655440000",
                "name": "edge-demo",
                "description": "Host + service on shared bridge",
                "status": "draft",
                "runtime_target": "docker",
                "networking_mode": "docker_bridge",
                "config": None,
                "created_at": "2025-01-15T10:00:00Z",
                "updated_at": "2025-01-15T10:00:00Z",
            }
        },
    )

    id: UUID
    name: str
    description: str | None
    status: TopologyStatus
    runtime_target: str
    networking_mode: str
    config: dict[str, Any] | None
    created_at: datetime
    updated_at: datetime


class TopologyNodeCreate(BaseModel):
    """Payload for adding a node to a topology graph."""

    name: str = Field(..., min_length=1, max_length=255)
    node_type: NodeType
    image: str | None = Field(default=None, max_length=512)
    ip_address: str | None = Field(default=None, max_length=64)
    config: dict[str, Any] | None = None


class TopologyNodeResponse(BaseModel):
    """Topology node returned from the API."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    topology_id: UUID
    name: str
    node_type: NodeType
    image: str | None
    ip_address: str | None
    config: dict[str, Any] | None


class TopologyLinkCreate(BaseModel):
    """Payload for adding a link between two nodes in the same topology."""

    source_node_id: UUID
    target_node_id: UUID
    network_name: str = Field(..., max_length=255)
    cidr: str | None = Field(default=None, max_length=64)
    gateway: str | None = Field(default=None, max_length=64)
    vlan_tag: int | None = Field(default=None, ge=0, le=4094)
    source_endpoint_ip: str | None = Field(default=None, max_length=64)
    target_endpoint_ip: str | None = Field(default=None, max_length=64)
    config: dict[str, Any] | None = None


class TopologyLinkResponse(BaseModel):
    """Topology link returned from the API."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    topology_id: UUID
    source_node_id: UUID
    target_node_id: UUID
    network_name: str
    cidr: str | None
    gateway: str | None = None
    vlan_tag: int | None = None
    source_endpoint_ip: str | None = None
    target_endpoint_ip: str | None = None
    config: dict[str, Any] | None = None


class TopologyUpdate(BaseModel):
    """Partial update for topology metadata."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    status: TopologyStatus | None = None
    runtime_target: str | None = Field(default=None, max_length=64)
    networking_mode: str | None = Field(default=None, max_length=64)
    config: dict[str, Any] | None = None


class TopologyNodeUpdate(BaseModel):
    """Partial update for a topology node (including merged config / UI positions)."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    node_type: NodeType | None = None
    image: str | None = Field(default=None, max_length=512)
    ip_address: str | None = Field(default=None, max_length=64)
    config: dict[str, Any] | None = None


class TopologyLinkUpdate(BaseModel):
    """Partial update for a topology link."""

    network_name: str | None = Field(default=None, max_length=255)
    cidr: str | None = Field(default=None, max_length=64)
    gateway: str | None = Field(default=None, max_length=64)
    vlan_tag: int | None = Field(default=None, ge=0, le=4094)
    source_endpoint_ip: str | None = Field(default=None, max_length=64)
    target_endpoint_ip: str | None = Field(default=None, max_length=64)
    config: dict[str, Any] | None = None

File: frontend/src/types/topology.ts
/** Topology graph intent (nodes, links) persisted by the control plane. */

export type TopologyStatus = 'draft' | 'active' | 'archived';

export type NodeType = 'generic' | 'router' | 'switch' | 'host' | 'gateway';

export interface TopologyResponse {
  id: string;
  name: string;
  description: string | null;
  status: TopologyStatus;
  runtime_target: string;
  networking_mode: string;
  config: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  /** Owning project when using multi-tenant projects. */
  project_id?: string | null;
  /** Present on list/detail API responses. */
  node_count?: number | null;
  link_count?: number | null;
}

export interface TopologyCreate {
  name: string;
  description?: string | null;
  runtime_target: string;
  networking_mode: string;
  status?: TopologyStatus | null;
  config?: Record<string, unknown> | null;
  /** When omitted, the API uses your default/first project. */
  project_id?: string | null;
}

export interface TopologyNodeResponse {
  id: string;
  topology_id: string;
  name: string;
  node_type: NodeType;
  image: string | null;
  ip_address: string | null;
  config: Record<string, unknown> | null;
}

export interface TopologyNodeCreate {
  name: string;
  node_type: NodeType;
  image?: string | null;
  ip_address?: string | null;
  config?: Record<string, unknown> | null;
}

export interface TopologyLinkResponse {
  id: string;
  topology_id: string;
  source_node_id: string;
  target_node_id: string;
  network_name: string;
  cidr: string | null;
  gateway?: string | null;
  vlan_tag?: number | null;
  source_endpoint_ip?: string | null;
  target_endpoint_ip?: string | null;
  config: Record<string, unknown> | null;
}

export interface TopologyLinkCreate {
  source_node_id: string;
  target_node_id: string;
  network_name: string;
  cidr?: string | null;
  gateway?: string | null;
  vlan_tag?: number | null;
  source_endpoint_ip?: string | null;
  target_endpoint_ip?: string | null;
  config?: Record<string, unknown> | null;
}

/** Persisted in `TopologyNode.config.editor_position` for React Flow layout. */
export interface EditorPosition {
  x: number;
  y: number;
}

export interface TopologyUpdate {
  name?: string;
  description?: string | null;
  status?: TopologyStatus;
  runtime_target?: string;
  networking_mode?: string;
  config?: Record<string, unknown> | null;
}

export interface TopologyNodeUpdate {
  name?: string;
  node_type?: NodeType;
  image?: string | null;
  ip_address?: string | null;
  config?: Record<string, unknown> | null;
}

export interface TopologyLinkUpdate {
  network_name?: string;
  cidr?: string | null;
  gateway?: string | null;
  vlan_tag?: number | null;
  source_endpoint_ip?: string | null;
  target_endpoint_ip?: string | null;
  config?: Record<string, unknown> | null;
}

export const EDITOR_POSITION_KEY = 'editor_position' as const;
